Improve readme with more guidance (#35)
Particular/PlatformContainerExamples: docker-compose/README.md

### Prerequisites

1. **SSL Certificate**: A PFX certificate file — see [Generating a certificate](#generating-a-certificate-local-testing-only) below
2. **CA Bundle**: A PEM file containing the CA that signed your certificate — see [Generating a CA bundle](#generating-a-ca-bundle-local-testing-only) below
3. **Microsoft Entra ID App Registration**: Configure an app registration for authentication

### Configuration

Update your `.env` file with the following values, replacing the `{placeholder}` values with your actual configuration:

> ```text

Reviewer: ```suggestion
```env
```

| `SERVICECONTROL_AUDIENCE` | Application ID URI from ServiceControl app registration (e.g., `api://{servicecontrol-client-id}`)                                       |
| `SERVICEPULSE_CLIENTID`   | Application (client) ID from ServicePulse app registration                                                                               |
| `SERVICEPULSE_APISCOPES`  | Array of API scopes ServicePulse should request when calling ServiceControl (e.g., `["api://{servicecontrol-client-id}/{scope-name}"]`)   |

#### Identity provider configuration examples

The `.env` values for `IDP_AUTHORITY`, `SERVICECONTROL_AUDIENCE`, `SERVICEPULSE_CLIENTID`, and `SERVICEPULSE_APISCOPES` depend on which identity provider you use.

<details>
<summary>Microsoft Entra ID</summary>

Register two app registrations in [Entra ID](https://entra.microsoft.com):

- **ServiceControl**: expose an API with an application ID URI and at least one scope
- **ServicePulse**: register as a single-page application; grant it permission to call the ServiceControl API

> ```text

Reviewer: ```suggestion
```env
```

IDP_AUTHORITY="https://login.microsoftonline.com/{tenant-id}"
SERVICECONTROL_AUDIENCE="api://{servicecontrol-client-id}"
SERVICEPULSE_CLIENTID="{servicepulse-client-id}"
SERVICEPULSE_APISCOPES=["api://{servicecontrol-client-id}/{scope-name}"]
```

</details>

<details>
<summary>Auth0</summary>

In your [Auth0 dashboard](https://manage.auth0.com):

- **API**: create an API whose *identifier* becomes `SERVICECONTROL_AUDIENCE`. Add a permission (scope) for ServicePulse to request.
- **Application**: create a Single Page Application for ServicePulse. Add `https://localhost:9090` to *Allowed Callback URLs*, *Allowed Logout URLs*, and *Allowed Web Origins*.

> ```text

Reviewer: ```suggestion
```env
```

IDP_AUTHORITY="https://{tenant}.auth0.com/"
SERVICECONTROL_AUDIENCE="{api-identifier}"
SERVICEPULSE_CLIENTID="{servicepulse-application-client-id}"
SERVICEPULSE_APISCOPES=["{api-identifier}/{permission-name}"]
```

> [!NOTE]
> The Auth0 authority URL requires a trailing slash.

</details>

<details>
<summary>Keycloak</summary>

In your Keycloak admin console:

- **ServiceControl client**: create a client with *Client authentication* enabled. Under *Client scopes*, add a dedicated scope (e.g., `servicecontrol:read`). Add an *Audience* mapper so the access token includes the ServiceControl client ID in the `aud` claim.
- **ServicePulse client**: create a public client with *Standard flow* and PKCE enabled. Add the ServiceControl scope as an optional or default scope.

> ```text

Reviewer: ```suggestion
```env
```

IDP_AUTHORITY="https://{keycloak-host}/realms/{realm}"
SERVICECONTROL_AUDIENCE="{servicecontrol-client-id}"
SERVICEPULSE_CLIENTID="{servicepulse-client-id}"
SERVICEPULSE_APISCOPES=["{servicecontrol-client-id}/{scope-name}"]
```

</details>

<details>
<summary>Duende IdentityServer</summary>

In your IdentityServer configuration:

- **ApiResource**: register a resource for ServiceControl with one or more scopes (e.g., `servicecontrol:read`).
- **Client**: register a client for ServicePulse using the `authorization_code` grant with PKCE. Grant it access to the ServiceControl scopes and set `RedirectUris` to `https://localhost:9090`.

> ```text

Reviewer: ```suggestion
```env
```

IDP_AUTHORITY="https://{your-identity-server}"
SERVICECONTROL_AUDIENCE="{api-resource-name}"
SERVICEPULSE_CLIENTID="{servicepulse-client-id}"
SERVICEPULSE_APISCOPES=["{api-resource-name}/{scope-name}"]
```

</details>

### Generating a certificate (local testing only)

> [!WARNING]
> The certificates generated below are for local testing only. Use a certificate from a trusted Certificate Authority for production deployments.

> [!IMPORTANT]
> The certificate must include every hostname used to access services over HTTPS. In a Docker Compose network, containers reach each other using their service names as hostnames (e.g., `https://servicecontrol:33333`). The TLS handshake checks that the server certificate contains a [Subject Alternative Name](https://en.wikipedia.org/wiki/Subject_Alternative_Name) (SAN) matching the hostname being connected to — if a name is missing, the connection is rejected.

#### Option 1: Using mkcert

> Install [mkcert](https://github.com/FiloSottile/mkcert):
> 
> ```pwsh
> winget install FiloSottile.mkcert
> ```

Reviewer: ```suggestion
Install [mkcert](https://github.com/FiloSottile/mkcert), installation instructions for Windows, Mac and Linux at:

- <https://github.com/FiloSottile/mkcert#installation>
```
